bevy_lint/src/lints/panicking_methods.rs

```diff
@@ -0,0 +1,290 @@
+        } else if match_type(cx, ty, &crate::paths::QUERY_STATE) {
+            Some(Self::QueryState)
+        } else if match_type(cx, ty, &crate::paths::WORLD) {
+            Some(Self::World)
+        } else {
+            None
+        }
+    }
+
+    /// Returns a list of panicking `Query` or `QueryState` methods and their non-panicking
+    /// alternatives.
+    ///
+    /// Each item in the returned [`slice`] is of the format
+    /// `(panicking_method, alternative_method)`.
+    fn alternatives(&self) -> &'static [(&'static str, &'static str)] {
+        match self {
+            Self::Query => &[
+                ("single", "get_single"),
+                ("single_mut", "get_single_mut"),
+                ("many", "get_many"),
+                ("many_mut", "get_many_mut"),
+            ],
+            Self::QueryState => &[
+                ("single", "get_single"),
+                ("single_mut", "get_single_mut"),
+                // `QueryState` does not currently have `many()` or `many_mut()`.
+            ],
+            Self::World => &[
+                ("entity", "get_entity"),
+                ("entity_mut", "get_entity_mut"),
+                ("many_entities", "get_many_entities"),
+                ("many_entities_mut", "get_many_entities_mut"),
+                ("resource", "get_resource"),
+                ("resource_mut", "get_resource_mut"),
+                ("resource_ref", "get_resource_ref"),
+                ("non_send_resource", "get_non_send_resource"),
+                ("non_send_resource_mut", "get_non_send_resource_mut"),
+                ("schedule_scope", "try_schedule_scope"),
+            ],
+        }
+    }
+
+    /// Returns the name of the type this variant represents.
+    fn name(&self) -> &'static str {
+        match self {
+            Self::Query => "Query",
+            Self::QueryState => "QueryState",
+            Self::World => "World",
+        }
+    }
+
+    /// Returns the [`Lint`] associated with this panicking type.
+    ///
+    /// This can either return [`PANICKING_QUERY_METHODS`] or [`PANICKING_WORLD_METHODS`].
+    fn lint(&self) -> &'static Lint {
+        match self {
+            Self::Query | Self::QueryState => PANICKING_QUERY_METHODS.lint,
+            Self::World => PANICKING_WORLD_METHODS.lint,
+        }
+    }
+}
+
+/// Returns the [`Span`] of an array of method arguments.
+///
+/// [`ExprKind::MethodCall`] does not provide a good method for extracting the [`Span`] of _just_
+/// the method's arguments. Instead, it contains a [`slice`] of [`Expr`]. This function tries it's
+/// best to find a span that contains all arguments from the passed [`slice`].
+///
+/// This function assumes that `args` is sorted by order of appearance. An [`Expr`] that appears
+/// earlier in the source code should appear earlier in the [`slice`].
+///
+/// If there are no [`Expr`]s in the [`slice`], this will return [`Span::default()`].
+fn span_args(args: &[Expr]) -> Span {
```

Comment:
This function sounds like a reusable helper function, which might also be useful to other lints.

Do you think it makes sense to start some kind of utility module for these kind of things?

Reply:
That's a good idea, but I'm going to defer that for a later PR. I think `span_args()` would need to be made less restrictive if it were to be shared across lints.